cpp/src/arrow/util/checked_cast.h

```diff
@@ -32,28 +32,28 @@ inline OutputType checked_cast(InputType&& value) {
   static_assert(std::is_class<typename std::remove_pointer<
                     typename std::remove_reference<OutputType>::type>::type>::value,
                 "checked_cast output type must be a class");
-#ifdef NDEBUG
```

Comment:
CI failed here because it explicitly undefined NDEBUG on a release build. I don't think these are public headers, so I could omit the change here and leave it as it was.